.pyinstaller/run_sunpy_tests.py

```diff
@@ -0,0 +1,4 @@
+import os
+import sys
+
+sys.exit(os.system('pytest -p no:warnings --doctest-rst -m "not mpl_image_compare" --pyargs sunpy'))
```

Comment:
No no no! This will just run regular unfrozen pytest so it's not testing anything except `os.system()`! Use `pytest.main()` instead.
```suggestion
import pytest

# Set some default args.
args = ['-p', 'no:warnings', '--doctest-rst', '-m',
        'not mpl_image_compare. '--pyargs sunpy']
# And allow additional arguments to be supplied at runtime.
args += sys.argv[1:]

pytest.main(args)
```

However, without an `import sunpy` in there, PyInstaller will have no reason to think that `sunpy` is a dependency so it won't collect it (or run the hook) unless you force it to with a `--hiddenimport=sunpy`. 

That's still probably not the end of the story though - it depends: Does `import sunpy` load everything so that you can access any public corner without additional import statements or are you expected to use `from sunpy.foo.bar import xyz` to pull the bits you need out? If it's the latter then PyInstaller will consider `sunpy.foo.bar` to be unneeded unless you add it as a `hiddenimport` or an explicit reference in `run_sunpy_tests.py` too. (This won't apply if you opt to keep the `collect_submodules("sunpy")` in the hook.)

It's quite likely that you'll have to have more than one test script, each importing and testing a different corner of `sunpy`, and each using only the optional dependencies that that corner requires.

Reply:
Thanks for your help on this @bwoodsend.